### PR TITLE
Apply AutoValue to Tag, TagKey, and TagValue.

### PR DESCRIPTION
**Pull request by sebright**

### CODE REVIEW

File: core/src/test/java/com/google/instrumentation/stats/TagTest.java [hunk 1]
```diff
@@ -52,9 +52,9 @@ public void testHashCode() {
 
   @Test
   public void testToString() {
-    assertThat(Tag.create(K1, V1).toString()).isEqualTo("Tag<k1,v1>");
-    assertThat(Tag.create(K2, V1).toString()).isEqualTo("Tag<k2,v1>");
-    assertThat(Tag.create(K1, V2).toString()).isEqualTo("Tag<k1,v2>");
+    assertThat(Tag.create(K1, V1).toString()).isEqualTo("Tag{key=k1, value=v1}");
+    assertThat(Tag.create(K2, V1).toString()).isEqualTo("Tag{key=k2, value=v1}");
+    assertThat(Tag.create(K1, V2).toString()).isEqualTo("Tag{key=k1, value=v2}");
```
Comment (sebright) [inline core/src/test/java/com/google/instrumentation/stats/TagTest.java]: Should we remove the toString() tests, now that we are using AutoValue?  Another option is to just test that the output of toString() is not the default, as suggested here: http://stackoverflow.com/a/12528424

File: core/src/main/java/com/google/instrumentation/stats/TagKey.java [hunk 2]
```diff
@@ -30,27 +35,14 @@
...
+    return new AutoValue_TagKey(StringUtil.sanitize(key));
   }
 
-  @Override
-  public int hashCode() {
-    return key.hashCode();
-  }
+  public abstract String asString();
 
+  // TODO(sebright) Use the default AutoValue toString(), which is more useful for debugging.
```
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/stats/TagKey.java]: Why do we not use that?

File: core/src/test/java/com/google/instrumentation/stats/TagTest.java (hunk 1 above)
Comment (bogdandrutu) [inline core/src/test/java/com/google/instrumentation/stats/TagTest.java]: I think if the class uses AutoValue we should do that.

File: core/src/main/java/com/google/instrumentation/stats/TagKey.java (hunk 2 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/stats/TagKey.java]: gRPC is currently calling toString() to access the key and value fields.  I added that to the comment.

File: core/src/test/java/com/google/instrumentation/stats/TagTest.java (hunk 1 above)
Comment (sebright) [inline core/src/test/java/com/google/instrumentation/stats/TagTest.java]: I removed it.

**Pull request merged**